### ISSUE TITLE
Support remote DOCKER_HOST by copying configs and secrets files instead of bind mounting them

### ISSUE THREAD

**Issue by andoks** [docker/compose#11867]
### Description

Per title, currently docker compose uses bind mounts to put configs and secrets in containers, but this does not work when using either docker context or DOCKER_HOST variable to remotely deploy to a different system.

This would help some of the use cases described in these issues:
- #7857 
- #11141
- #5523
- #7110

It builds upon the config content and environment support added in https://github.com/compose-spec/compose-spec/issues/346

**Comment by andoks** [comment]
A suggestion for how to fix this, see the following attached patch that should apply to main branch
[0001-Enable-config.file-s-on-remote-docker-hosts.patch.gz](https://github.com/user-attachments/files/15506646/0001-Enable-config.file-s-on-remote-docker-hosts.patch.gz) (had to gzip the file, as it would not upload as a plain .patch file)

**Comment by andoks** [comment]
Two alternatives if this change in existing behaviour is undesireable, could be to make the config.driver = "copy" change to this new behaviour, or support a new config.copy property for copying the file into the container/service.

@ndeloof hope it is ok I am tagging you here, as I saw you were involved in the https://github.com/compose-spec/compose-spec/issues/346 issue

**Comment by ndeloof** [comment]
I fully agree with this direction.
Regarding change in behavior, I don't feel like this is an issue as long as this is documented in release note.
A trivial workaround for those who require dynamic sync with container would be to declare an additional bind mount or to rely on a dedicated `watch` trigger to sync changes

**Comment by ndeloof** [comment]
> see the following attached patch

Please open a pull-request

**Comment by ndeloof** [comment]
related: https://github.com/docker/compose/issues/9648#issuecomment-1557068598

**Comment by andoks** [comment]
> > see the following attached patch
> 
> Please open a pull-request

@ndeloof: done -> https://github.com/docker/compose/pull/11871

**Comment by LawrenceWarren** [comment]
This is a good feature that lends itself to a GitOps approach to remote Docker Compose deployments. However I think this should be a toggle, especially with secrets. 

For example, in my workflow I have various `docker-compose.yml`s each in different directories, with each directory corresponding to a remote server. As a very basic example:

```tree
infrastructure-as-code
├── dev
│   └── instances
│       ├── dev-server-1
│       │   └── docker-compose.yml
│       └── dev-server-2
│           └── docker-compose.yml
└── prod
    └── instances
        ├── prod-server-1
        │   └── docker-compose.yml
        └── prod-server-2
            └── docker-compose.yml
```

This is all Git tracked as a single source of truth to track server state and changes to deployments over time. By being able to use configs from my localhost, we can also track Docker config changes in Git. However, it doesn't make sense to track secrets in the Git repository, as they are meant to be secret - it would be more ideal for my workflow to bind/copy secrets from DOCKER_HOST.

I think an attribute such as `from_docker_host` would be useful, so you can explicitly choose where the config/secret files will be found when using Docker contexts, and it is flexible around different workflows people may use:

```yaml
configs:
  my_config:
    file: ./service.conf
    from_docker_host: false # Will search my localhost for service.conf

secrets:
  my_secret:
    file: /etc/my-docker-secerts/private.pem
    from_docker_host: true # Will search DOCKER_HOST for /etc/my-docker-secerts/private.pem
```

**Comment by LawrenceWarren** [comment]
@andoks - You're probably right, my idea/request may be outside the scope of this specific issue. I think in general secrets & configs are not quite perfect when it comes to interactions with Docker contexts. Before creating my own issue to discuss this, I searched GitHub for a similar issue and found this.

To your question, yes, my example with secrets does work, and I do currently use it to securely write secrets into containers. In my opinion the current behaviour (reading the file path as a path on DOCKER_HOST, and mounting that file) has the correct effect when it comes to secrets, and this merge request would break this workflow.

**Comment by ndeloof** [comment]
@LawrenceWarren that's an interesting use-case that demonstrate we can't adopt this feature without considerations for the impacts on existing installations. Bind mounts indeed always target a path on remote Docker Host

**Comment by LawrenceWarren** [comment]
@ndeloof Switching to copying does not necessarily break the workflow mentioned above, so long as you're still copying from DOCKER_HOST.

Of course, the point of this issue is to use copying so that we can source data from localhost, but that is a breaking change if we make it the default behaviour. I am unsure how many people are using my workflow, but we must assume that there are people out there using these features together in their current format, particularly with secrets, since (in my opinion) reading from DOCKER_HOST is probably the desired effect.

I think having a key to specify where the file is found is a simple solution (and the default should be `from_docker_host: true` so that existing config does not suddenly have different behaviour), but this expands the scope of the issue and pull request significantly.

**Comment by ndeloof** [comment]
> so long as you're still copying from DOCKER_HOST.

Which is the blocker here: Compose being a client-side command, it can't access files on DOCKER_HOST, can only ask daemon to set a bind mount.

**Comment by andoks** [comment]
@ndeloof: what do you think about introducing a new key (e.g "copy_file", "local_file" or something else that makes sense) that can do it the new way? Perhaps by setting "driver: copy" maybe? I mostly open for anything as long as it enables use of local (configuration) files while deploying remotely. I assume this may also affect the compose spec, which requires a somewhat holistic view

**Comment by andoks** [comment]
@LawrenceWarren: is there any particular reason for why you use the config stanza instead of bind-mounting your secret explicitly? I would think it could be reasonable to go through with this change if bind-mounting explicitly works, as that seems more in line with the use case you specified above?

**Comment by ndeloof** [comment]
`secrets` abstraction should be preferred over bind mount as this clarify role of mounts in container. Also will allow to benefit native support for secrets if/once docker engine introduce support for those (even when not running in swarm mode)

**Comment by LawrenceWarren** [comment]
@andoks I use the secrets stanza for our secrets because... that's what it is for, I suppose. You're right, they could be explicitly mounted as a volume, but this also seems bad as again it would be a breaking change. We can't assume there are no users out there where this wouldn't break things.

**Comment by andoks** [comment]
> secrets abstraction should be preferred over bind mount as this clarify role of mounts in container. Also will allow to benefit native support for secrets if/once docker engine introduce support for those (even when not running in swarm mode)

I agree that using the secrets feature to define secrets makes sense, but the secrets.file (and configs.file) behavior should be well defined should it not? Either it always refers to a local file that resides on the filesystem together with the compose file (the changed behavior), or it is defined to exist on the file system where the deployment is being run (the current implementation). And I guess that should perhaps be included in the spec somehow? 

If it is the former, then applying the change, possibly with a new keyword (some suggestions above) for retaining the current functionality could be made, but break existing users until they update to the new method. 

If it is the latter, then the current implementation should be retained, and if desirable by upstream, I can change my PR to work with a different configs/secrets property (new configs.<property> property, setting the driver or something else) if it is acceptable to support both ways.

Also, I take it from your comment that docker in swarm mode has some engine-level primitive for secrets that the current compose implementation tries to mimic (currently by bind mounts)?

I guess this might warrant some more input from other docker maintainers?

I will follow the issue, and if it is decided on, I can try to change my implementation accordingly :slightly_smiling_face: 

From https://docs.docker.com/compose/use-secrets/ I read it as the files used are local to the docker-compose.yml filesystem since the files are specified relative, but it does not call out anything explicitly (neither does https://docs.docker.com/compose/compose-file/09-secrets/ AFAICS)

**Comment by ndeloof** [comment]
By allowing paths to be relative, Compose always made the assumption docker daemon is local. This obviously isn't always the case, still this is obvious in the compose model in many places, including here when it comes to accessing secrets/configs files

**Comment by andoks** [comment]
@ndeloof: do you know what is needed to arrive at a decision on this to move forward?

**Comment by LawrenceWarren** [comment]
Hi, I hope everyone had a good weekend, just coming in to bump the topic a bit.

I found this issue last week while trying to see if there was already an issue open about writing configs from localhost when DOCKER_HOST is set. While I have thrown a spanner in the works by mentioning my secrets use case, I would still like to see configs from localhost as a viable possibility.

If there were a way to create configs from localhost, while retaining the current behaviour of secrets (and therefore kicking the issue of "native support for secrets" that @ndeloof mentioned down the road) I would be grateful. Docker contexts is a brilliant tool that is 90% of the way there, it would be good to see these changes made.

**Comment by andoks** [comment]
> Hi, I hope everyone had a good weekend...

Thanks, same to you :-)

> ..., just coming in to bump the topic a bit.

> I found this issue last week while trying to see if there was already an issue open about writing configs from localhost when DOCKER_HOST is set. While I have thrown a spanner in the works by mentioning my secrets use case, I would still like to see configs from localhost as a viable possibility.

I realize this might be a figure of speech with regards to the spanner, but I do not think that is the case at all. As with any widely used open source project like this, you are probably just one out of potentially 10's/100's/1000's that use the configs.file support with the expectation of docker binding the file from a potential remote host, which is very useful to know. At least this way, even if the behavior is changed, the project knows that it will be important to at least document this potentially breaking users, and potentially supply descriptions of workarounds ahead of time for anyone that are affected by this change.
 
> If there were a way to create configs from localhost, while retaining the current behaviour of secrets (and therefore kicking the issue of "native support for secrets" that @ndeloof mentioned down the road) I would be grateful. Docker contexts is a brilliant tool that is 90% of the way there, it would be good to see these changes made.

My 2 cents regarding this: I do not think it is a good idea to make the behavior of secrets and configs differ for consistency, but also with an expectation that there might be users that will be broken by the change to configs anyway, using it to bind configs on remote hosts today.

**Comment by andoks** [comment]
@ndeloof: sorry to bother you again if you are busy or otherwise preoccupied, but do you know what is needed to move forward with this?

**Comment by ndeloof** [comment]
time ;)

**Comment by schaubl** [comment]
Hello @andoks @ndeloof @LawrenceWarren and others,

I'm the author of the PR #11984 which I just realized is a duplicate of #11871 (don't know why I missed it).

I've read carefully all the comments in this issue and below are my 2 cents:

I think the initial implementation of (file-based) configs&secrets in compose was already a kind of a breaking change compared to the implementation in swarm/stack (and the doc). File-based configs&secrets are also discrepant with inlined/env-based configs&secrets.

> ...
There's no technical reason docker engine can't offer secrets/configs natively, just this is guarded by Swarm mode, but AFAIK some discussion happened to get them also available in standalone mode. I can't tell if/when this would take place, but preferably the logic here should consider this may happen
...
Source: https://github.com/docker/compose/pull/11931#issuecomment-2219642915

I agree that it would totally make sense to be able to use docker configs&secrets when swarm mode is not enabled.
I think one of the reasons they are not available without swarm mode is because they are made to be distributed across swarm nodes. It probably also needs some swarm keys/certificates to decrypt secrets.
Considering it could be possible in the future to use docker's native configs&secrets, the source file should be read by the docker compose client because this is how it is implemented in docker swarm/stack and it's not possible AFAIK to ask the docker daemon to create a config/secret from the filesystem where the docker daemon reside.

> @LawrenceWarren: is there any particular reason for why you use the secrets stanza instead of bind-mounting your secret explicitly? I would think it could be reasonable to go through with this change if bind-mounting explicitly works, as that seems more in line with the use case you specified above?
...
Source: https://github.com/docker/compose/issues/11867#issuecomment-2147927511

Using the current implementation of docker compose secrets is effectively the same as declaring a bind mount.
It might give the impression that it adds some security/secrecy by using file-based secrets compared to a standard bind mount but in fact it doesn't.

The proposed change is a breaking change that "fix" the initial breaking change.

My personal opinion is that this change could be done with a proper mention in the release notes explaining how to use bind mounts to migrate in case file-based configs/secrets were used "remotely" (as mentioned [here](https://github.com/docker/compose/issues/11867#issuecomment-2141310118)).

**Comment by LawrenceWarren** [comment]
OK, I'm inclined to agree with you @schaubl because I think being able to copy files from your localhost, even when dockerhost is a different server, is the correct behaviour, and you are right that the same behaviour can be achieved using bind mounts.

However I would be cautious; just because this breaking change is technically a fix (as you put it), it could still break production workflows - in fact, it would break mine. Also, as mentioned above, the secrets stanza is a nice abstraction even if it fundamentally is just the same as a bind mount, so using a bind mount for secrets is definitely a step back.

**Comment by LaXiS96** [comment]
I also am in dire need of this feature, as my use case is a Windows development environment and multiple remote Linux Engines.

I feel the client-server architecture of Docker implies this kind of behavior, but it looks like many features expect the daemon to be running locally, which is unintuitive.

**Comment by domsew** [comment]
I would also like to support @schaubl opinion. Important feature of the `compose` is that we can create reproducible workspaces, that anyone can use without special host machine configurations. Ability to use workspace config / secret files is the last missing block which would make compose agnostic regarding the `docker context`. For now I have to switch somehow from relative to absolute path depending on context used (and push required files with a different path). Of course such versatilityt can be achieved now with `secrets.environment`, but it is not super convinient. It also ensures consistency with different compose attributes - enviroment variables, dockerfiles, build contexts - all of these are read locally from current workspace, not remotely. Configs and Secrets should work the same way.

I believe current behaviour should be changed. Current production workflows will be able to be simplified (no need to manually copy files), or straightforward migration to `volumes` exists.

**Comment by andoks** [comment]
For anyone interested in testing my fix, feel free to pull down the changes in #12251, building is quite straightforward. I'm already using it locally for my own convenience with regards to using file configs and secrets =)

**Comment by schaubl** [comment]
I recreated my [previous PR](https://github.com/docker/compose/pull/11984) (which got stalled) but based on the latest code base: 
- #12448

The idea is not to compete with @andoks PR (#12251) but I have a slightly different implementation which might participate to the final implementation and because #12251 is set as Draft and there was no actions made since 2 months.

It differs in two main points:
1. It checks the validity of the config/secret parameters in `getCreateConfigs` before the container is created where the other PR does it after.
2. It simply reuses the existing `createTar` function but doesn't support directories which correspond to the [doc](https://docs.docker.com/reference/compose-file/configs/) and the behaviour of Docker [Configs](https://docs.docker.com/reference/cli/docker/config/create/)&[Secrets](https://docs.docker.com/reference/cli/docker/secret/create/) used by Swarm/Stack.

I personally don't have strong opinions if directories should be supported or not, I think it depends if the idea is to stick to the implementation of `docker stack` or to provide extra features. 

However I really think files for config&secrets should come from the local host and not bind mounted (bind mounts can be used for that) for all the reasons mentioned in this thread.

@ndeloof If there is anything I can do to help this topic to progress, please let me know and I will be happy to do so.

**Comment by MikaelElkiaer** [comment]
This is on the very top of my wish list.
I have several scenarios where this would make sense, e.g.:

1. Wrapping `docker compose` usage in another container - which could carry out bootstrapping and then use the host socket for Compose commands.
2. Hosting a stack in a pod - having a sidecar with the daemon.

Any idea when this will be resolved?
Or whether it will at all?

**Comment by ndeloof** [comment]
@MikaelElkiaer I wish it will be, but AFAICT this will require some dedicated syntax to "opt-in" for copy vs being the default implementation, as this has some non-trivial impacts

**Comment by MikaelElkiaer** [comment]
Could it be done via an option - i.e. env var or CLI argument?

I could see this as something where you either want all secrets/configs to be either mounted or copied and not necessarily on an individual level.
You would use an existing stack on your local host simply as is.
Then if you want to use an existing stack with a remote host you could set the `--remote-copy=true` arg or `COMPOSE_REMOTE_COPY=1` env var or some such.

**Comment by ndeloof** [comment]
> we are kind of in a limbo anyway, where the swarm and compose implementations works differently, which is also not good I assume.

Exact. That's a major issue. `secrets` and `config` have been introduced in compose file format for Docker Swarm but are not supported by standalone Docker Engine, and as such Compose only can offer a hack to pretend those are also supported, with many limitations / side-effects / frustration. Using `docker cp` to inject secrets in container unlocked some users but introduced some new issues and challenges we haven't yet addressed.

>  what would be the way forward to resolve this issue

I can't tell. I'm :-1 introducing even more hacks in this area to pretend compose can manage secrets, with risks for sensitive data leaks / more weird bugs / confusing behaviors - typically, some rely on container to get secrets updated after a local change as using bind-mount suggest things _should_ happen, while this is undocumented and sometime not desired.

**Comment by andoks** [comment]
> > we are kind of in a limbo anyway, where the swarm and compose implementations works differently, which is also not good I assume.
> 
> Exact. That's a major issue. `secrets` and `config` have been introduced in compose file format for Docker Swarm but are not supported by standalone Docker Engine, and as such Compose only can offer a hack to pretend those are also supported, with many limitations / side-effects / frustration. Using `docker cp` to inject secrets in container unlocked some users but introduced some new issues and challenges we haven't yet addressed.
> 
> > what would be the way forward to resolve this issue
> 
> I can't tell. I'm :-1 introducing even more hacks in this area to pretend compose can manage secrets, with risks for sensitive data leaks / more weird bugs / confusing behaviors - typically, some rely on container to get secrets updated after a local change as using bind-mount suggest things _should_ happen, while this is undocumented and sometime not desired.

Assuming you work for Docker (the company), is there anyone we could loop into the discussion to find a way forward, or is it perhaps also an issue of priorities, where they won't have the time to dedicate to this issue?

(feel free to ignore this paragraph, it's just me restating my previous arguments for why I think it is better to change the current behaviour of config stanzas)
And although I agree and see all of your points, about backwards compatibility, I (personally) still am leaning towards that if nothing else, the config-stanzas could be changed to copy the contents into the container rather than mounting them, simply for the reason that AFAIK, anyone that needs the bind-mount behaviour, should be able to actually use bind-mount directly to get the existing behaviour, while the rest of us do not have any way of injecting configuration that works transparently over the network within the confines of the docker compose file.

**Comment by alexey-sh** [comment]
One can only conclude that this problem is doomed to persist indefinitely, a natural consequence of docker woefully misguided design.

**Comment by mrtrizer** [comment]
I'm new to docker and everything looked nice and shiny when I was setting everything up according to the docs and everything was working perfectly locally.
But then I got a misleading error about absolute path error.
While this problem persists, is there any workaround for this? Without baking the secrets into the images, like I did in the past.
Maybe there is an option of building containers with volumes locally and uploading them? Or should I copy everything to the server and run compose there?

**Comment by ndeloof** [comment]
@mrtrizer copy sensible file to your remote server, then set compose to use an absolute path:
```yaml
secrets:
  foo:
    file: /absolute/path/on/remote/server
```

You also can just define a bind mount for your service, as this is actually how things eventually take place:

```yaml
services:
  foo:
     volumes:
       /absolute/path/on/remote/server:/run/secrets/foo
```

**Comment by mrtrizer** [comment]
@ndeloof Thank you, this should do as a workaround. But unfortunately, in this case with absolute paths, I can't test it locally. Is there a way to use relative paths and pass some path of the remote machine as an argument in the compose cli?
So I could scp ./secrets to path/on/remote and then just do something like compose up --root=path/on/remote

**Comment by ndeloof** [comment]
@mrtrizer use an override compose file for local development, to switch remote path to a local one

**Comment by andoks** [comment]
@mrtrizer 
> I'm new to docker and everything looked nice and shiny when I was setting everything up according to the docs and everything was working perfectly locally. But then I got a misleading error about absolute path error. While this problem persists, is there any workaround for this? Without baking the secrets into the images, like I did in the past. Maybe there is an option of building containers with volumes locally and uploading them? Or should I copy everything to the server and run compose there?

If the content of the files are not too large, you can also consider using the [content](https://docs.docker.com/reference/compose-file/configs/#example-2) stanza instead of the file stanza, this way you can test it locally without any work-arounds (except for using content instead of file obviously)

**Comment by mrtrizer** [comment]
In the end I came to this solution. I'm not sure it's entirely secure, but good enough for my application. Use environment variable with empty default value and use the same variable to copy the files to the server. Not much overhead and works almost as I expected it to be working by default with docker compose.

```yaml
secrets:
  foo:
    file: ${SECRETS_DIR:-}./secrets/foo.json
```

```powershell
$env:SECRETS_DIR = "/etc/docker/"

scp -r "$($env:LOCAL_SECRETS_DIR)/*" "$($env:REMOTE_HOST):$($env:SECRETS_DIR)$($env:LOCAL_SECRETS_DIR)"

docker-compose --context remote -f $composeFile up --build -d $args
```

**Comment by scratchmex** [comment]
I think reading the localhost file if the source path is relative **and** the docker host is remote (not unix or win socket). Would that work for backwards compatibility? I think that behavior is broken today so it would be a feature that does not affect current working behavior

**Comment by scratchmex** [comment]
this has been asked ten years ago here
- https://github.com/docker/compose/issues/1664

**Comment by scratchmex** [comment]
gentle ping enumerating all the PRs that were created to fix this and to discuss why they have not progressed, and sadly closed by inactivity

- https://github.com/docker/compose/pull/12448
- https://github.com/docker/compose/pull/12251
- https://github.com/docker/compose/pull/12625 (partial)

I volunteer to help in the coding necessary as well

from the maintainers perspective (@ndeloof @glours  @docker/compose-maintainers), what's stopping these feature? apart from time, which not a problem because we have a lot of interested persons able to code it, just the review part will be needed

cc @andoks @schaubl are you still interested in continuing the work?

**Comment by JojOatXGME** [comment]
I also think the current behavior of `{secrets|configs}.*.file` should be changed for multiple reasons:

* It is inconsistent with `{secrets|configs}.*.environment`
* It is allegedly incompatible with the reference implementation of Docker Swarm[^1]
* It doesn't make any sense with relative paths
* It prevents people from having a reasonable way to reference config files from their `docker-compose.yml`

> I have several scenarios where this would make sense, [...] 1. Wrapping `docker compose` usage in another container - which could carry out bootstrapping and then use the host socket for Compose commands.

This is what Portainer is doing and what brought me here. It makes it impossible to use any relative paths in compose files.

#### How to Fix

I think the following changes would be possible without causing too many issues with backwards compatibility:

* **Use local file when the path is relative.**
* **Keep using remote file when the path is absolute, but generate a warning.**
* Generate a warning even when the Docker Engine is running locally. Uses should be informed about the inconsistency as early as possible.
* Optionally add a command line option to keep using a bind-mount. Maybe this could be helpful for some people during local development.

Most people who want to upload local files are probably using relative paths anyway. They are probably talking about files inside the same directory/repository. At the same time, people who have used remote files for secrets have probably used absolute paths. After all, trying to access the path of your current working directory on the remote server doesn't make sense.

While the goal is to also change the behavior for absolute paths eventually, I think we could keep this implementation for decades. I think being able to use absolute paths is not that important in this context. Note that the warning means we are effectively discouraging absolute paths for `secrets` and `configs` altogether.

I would not introduce a new keyword.

#### About finding a long-term solution for secrets on the remote host

> since (in my opinion) reading from DOCKER_HOST is probably the desired effect.

@LawrenceWarren I don't know. I definetly would not have expected that. Also the [offical documentation of `secrets`](https://docs.docker.com/reference/compose-file/secrets/) uses a relative path, which doesn't make sense when we resolve the path on the remote.

> I use the secrets stanza for our secrets because... that's what it is for, I suppose.

I agree that in some scenarios you might want to reference secrets which are already stored remotely. I also agree that it would be cleaner if you could do it via the `secrets` section. However, I was wondering how Docker Swarm handles such scenarios. I would expect to have an option `external: true`, but [according to the spec](https://docs.docker.com/reference/compose-file/secrets/), this option only exists for `configs`. Unfortunately, I don't have any experience with Docker Swarm, which might be important to come up with a holistic solution. Anyway, here is my intuition:

* And an option `secrets.*.external` to the Compose specification.
* Implement support for `external: true` in Docker Compose by using volumes.

I think a straightforward and intuitive solution to implement `external: true` is to map them to volumes. Just search for a volume with a specific naming pattern. You should also verify the labels on the volume, which should indicate that the volume in indeed holding a secret. You may also add a command to Docker compose to upload secrets. If Docker ever introduces a native solution for secrets, Docker Compose could transparently switch to the new mechanism at any point.

Also note that even when you are not using `external: true`, it would probably be more appropriate to upload secrets and configs into a separate volume. After all, these files are supposed to be mounted read-only.

> `secrets` abstraction should be preferred over bind mount as this clarify role of mounts in container. Also will allow to benefit native support for secrets if/once docker engine introduce support for those (even when not running in swarm mode)

@ndeloof I agree, but based on the comments in this issue, the current implementation fails to keep compatibility with Docker Swarm, which might be even more problematic than not automatically getting the advanced security if Docker happens to ever implement native support for secrets. Anyway, my proposal above should resolve this concern as well.

[^1]: https://github.com/docker/compose/issues/11867#issuecomment-2232745553

**Comment by ndeloof** [comment]
> It is allegedly incompatible with the reference implementation of Docker Swarm[1]

Swarm is not reference implementation for the compose specification, it just adopted the compose file format at some point in time but didn't adopted recent updates.

> I was wondering how Docker Swarm handles such scenarios

Swarm relies on engine to manage secrets and config as an actual resource. Standalone Docker engine doesn't have such a feature.

> How to Fix

Your proposal sounds reasonable but devil is in the details. Any attempt we made to change this behavior resulted into many users to report issues with various corner cases. Typically, if we copy secret into container the same way we do for `environment`, changes to local file won't be reflected in container and we have no (simple) way to detect this. This would trigger some mismatch and confusion for users.

I can see two options here
1. Engine introduces support for secrets/configs. This has been discussed many time, as `build` would also benefits from such a feature
2. we introduce an explicit `copy` pre_start hook, so one can explicitly ask compose to inject a config/secret in container.

**Comment by JojOatXGME** [comment]
> Swarm is not reference implementation for the compose specification

I would agree that “reference implementation” is not the right word. However, based on the discussion above, my understanding was that the `secrets` and `configs` were introduced for Swarm, and Docker Compose followed suite for compatibility.

I would also say that the implementation of Docker Swarm (as described previously) seems quite reasonable, considering that secrets are often injected on the "build agent" during the deployment. This means you would have the secret available where you process the Compose file, but not necessarily on the host running the Docker Engine. The same applies to `configs`. It wouldn't surprise me if these options were intentionally implemented to read files from the "build agent", and not from `DOCKER_HOST`. However, I can only guess about intentions. I might be completely wrong.

> Swarm relies on engine to manage secrets and config as an actual resource.

Which I think could be mimicked very closely by using volumes.

>> I was wondering how Docker Swarm handles such scenarios

However, I actually meant that I don't know how the Compose file for Docker Swarm would look like for such use case. My intuition would have been that this is the purpose of the `external` flag, but it doesn't seem to exist for secrets?

> changes to local file won't be reflected in container

That is correct. This is why I suggested adding a flag to keep this behavior. But yes, it would still be a breaking change which needs to be well documented. I just think that it is a reasonable change to make. If we could somehow distinguish production scenarios from development scenarios with a local engine, we could even use different defaults depending on the scenario.

> I can see two options here [...] Engine introduces support for secrets/configs. [...]

In this case, I would assume switching to native secrets and configs from the Docker Engine would also break existing behavior in the same way. However, if Docker adding native support is on the horizon, then it makes sense to wait for that, of course.

**Comment by divStar** [comment]
I'd also be interested in this feature since I am running a Docker VM (Flatcar) and connecting to it using Docker. I don't want to manually pre-process the configuration files in order to copy them into the "content" property. I'd much rather like to specify the "file" (in the current folder of the `docker-compose.yml` for example) and e.g. state, that it should be loaded `from_local: true` or `from_host: false` or whatever.

As matters currently are, I am forced to pre-process this using whatever means, which isn't really helpful, because I want to keep my stack / container set up (I am not using Docker Swarm) as simple as possible.

**Comment by ndeloof** [comment]
Suggested workaround:

```yaml
services:
  test:
    build: 
       dockerfile_inline:
          FROM <service image>
          COPY config /etc/config
```
